### ISSUE TITLE
Tag Device In TensorGraph

### ISSUE THREAD

**Issue by lilleswing** [deepchem/deepchem#681]
We are starting to deploy DeepChem in server environments with multiple GPU's per node.  We wouldn't have to do this if our server environment implement proper cgroups or containerization -- but alas I think this will be standard across the industry.

I want to add a constructor arg for TensorGraph for the device so it won't try to take up a GPU that is supposed to be doing something else (say a DFT calculation).

This will not help with multi-gpu training for models, but is this actually a problem we realistically expect to have in the next year?

**Comment by rbharath** [comment]
Multiple GPUs per node seems likely to be an industry standard. I like the idea of being able to set a device in constructor.

I don't think we have immediate needs for multi GPU training. We should keep an eye on it, but for now seems like it's not yet worth the effort?

**Comment by peastman** [comment]
Let's not do anything that precludes using multiple GPUs at once.  Nvidia just gave us a pair of GPUs and nvlink connector specifically so we could work on that.

**Comment by lilleswing** [comment]
Well yeah kinda the point of this post -- how/are we planning on multi-gpu per model?  What would the API look like?  Can we make that API also satisfy tagginga whole TG model to a single gpu?  Do we have to do it on a per layer basis?

**Comment by lilleswing** [comment]
UPDATE:
We solved this via setting CUDA_VISIBLE_DEVICES environment variable for the python process rather than solving it in application logic.

So we no longer have a short term need to do this -- instead we should use this issue for tracking multi-gpu usage in the future.

**Comment by rbharath** [comment]
I've been looking into multi-GPU training a bit recently, using https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py

It makes a nice difference to be able to use a 2 GPU box effectively. I think we should start charting out what multi-GPU TensorGraph support might look like. Naively, I think this should be quite feasible. We would have to create a separate instance of the graph on each GPU node and train accordingly.

**Comment by simonfqy** [comment]
Hi all, have you made any progress in the GPU parallel processing? I am using WeaveTensorGraph and I could use 2 to 4 GPUs in our lab, but it seems only 1 GPU is actively working (with memory fully occupied at the same time) while all other GPUs have their memory fully occupied but not doing any work. I checked the issues on GitHub and it seems that this problem persists.

**Comment by rbharath** [comment]
Unfortunately, this looks to be a major undertaking. We are working on better interoperability between TensorGraph and raw TensorFlow, so you should be eventually able to use TensorFlow's distributed training support with DeepChem layers. Might take us a few months though since there are a lot of complicated inter-op questions to work out.

**Comment by simonfqy** [comment]
@rbharath Thanks for your reply. While your team is still working on the issue, I hope that you keep as much backward compatibility as possible. Currently I am using the deepchem library and inheriting lots of classes from it to create customized classes and member functions. If you go through some major upgrades and discard too much existing code, I am sure my code will break and it will take me a long time to learn the updated deepchem library. Backward compatibility should benefit the community a great deal.

**Comment by rbharath** [comment]
@simonfqy Yes, agreed that backwards compatibility is important. We're moving carefully to make sure that we avoid breaking existing code to the extent possible.

**Comment by rbharath** [comment]
Closing this issue. Please re-open if necessary.
